pkg/bindings/connection.go

```diff
@@ -17,33 +25,90 @@ type APIResponse struct {
-	if len(url) < 1 {
-		url = defaultConnection
+// NewConnection takes a URI as a string and returns a context with the
+// Connection embedded as a value.  This context needs to be passed to each
+// endpoint to work correctly.
+//
+// A valid URI connection should be scheme://
+// For example tcp://localhost:<port>
+// or unix://run/podman/podman.sock
+func NewConnection(uri string) (context.Context, error) {
+	u, err := url.Parse(uri)
+	if err != nil {
+		return nil, err
 	}
-	newConn := Connection{
-		url:    url,
-		client: &http.Client{},
+	// TODO once ssh is implemented, remove this block and
+	// add it to the conditional beneath it
+	if u.Scheme == "ssh" {
+		return nil, ErrNotImplemented
+	}
+	if u.Scheme != "tcp" && u.Scheme != "unix" {
+		return nil, errors.Errorf("%s is not a support schema", u.Scheme)
+	}
+
+	if u.Scheme == "tcp" && !strings.HasPrefix(uri, "tcp://") {
+		return nil, errors.New("tcp URIs should begin with tcp://")
 	}
-	response, err := http.Get(pingConnection)
+
+	address := u.Path
+	if u.Scheme == "tcp" {
+		address = u.Host
+	}
+	newConn := newConnection(u.Scheme, address)
+	ctx := context.WithValue(context.Background(), "conn", newConn)
+	if err := pingNewConnection(ctx); err != nil {
+		return nil, err
+	}
+	return ctx, nil
+}
+
+// pingNewConnection pings to make sure the RESTFUL service is up
+// and running. it should only be used where initializing a connection
+func pingNewConnection(ctx context.Context) error {
+	conn, err := GetConnectionFromContext(ctx)
 	if err != nil {
-		return newConn, err
+		return err
 	}
-	if err := response.Body.Close(); err != nil {
-		return newConn, err
+	// the ping endpoint sits at / in this case
+	response, err := conn.NewRequest(http.MethodGet, "../../../_ping", nil, nil)
+	if err != nil {
+		return err
 	}
-	return newConn, err
+	if response.StatusCode == http.StatusOK {
+		return nil
+	}
+	return errors.Errorf("ping response was %q", response.StatusCode)
+}
+
+// newConnection takes a scheme and address and creates a connection from it
+func newConnection(scheme, address string) Connection {
+	client := http.Client{
+		Transport: &http.Transport{
+			DialContext: func(_ context.Context, _, _ string) (net.Conn, error) {
+				return net.Dial(scheme, address)
+			},
+		},
+	}
+	newConn := Connection{
+		client:  &client,
+		address: address,
+		scheme:  scheme,
+	}
+	return newConn
 }
 
 func (c Connection) makeEndpoint(u string) string {
-	return fmt.Sprintf("%s%s", defaultConnection, u)
+	// The d character in the url is discarded and is meaningless
+	return fmt.Sprintf("http://d/%s%s", defaultConnectionPath, u)
 }
 
-func (c Connection) newRequest(httpMethod, endpoint string, httpBody io.Reader, params map[string]string) (*APIResponse, error) {
+// NewRequest assembles the http request and returns the response
+func (c Connection) NewRequest(httpMethod, endpoint string, httpBody io.Reader, params map[string]string) (*APIResponse, error) {
```

Comment:
This should be named DoRequest() or something like that.  It doesn't return a new API Request object().

If this is changed to a variadic function we could hide more boilerplate:

conn.NewRequest(nil, params, "GET", "/images/%s/json", nameOrID)

NewRequest(r io.Reader, p map[string]string, m string, e string, a ...interface{})
  s := fmt.Sprintf(e, ...a)
  // Use s
  .